bridge/nimbus/whisper.go

```diff
@@ -0,0 +1,377 @@
+
+		// Allocate a buffer for Nimbus to return the symkey on
+		dataC := C.malloc(C.size_t(whispertypes.AesKeyLength))
+		defer C.free(unsafe.Pointer(dataC))
+		if ok := C.nimbus_get_symkey(idC, (*C.uchar)(unsafe.Pointer(dataC))); !ok {
+			c <- errors.New("symkey not found")
+		}
+
+		c <- C.GoBytes(dataC, whispertypes.AesKeyLength)
+	})
+	if err, ok := retVal.(error); ok {
+		return nil, err
+	}
+
+	return retVal.([]byte), nil
+}
+
+//export onMessageHandler
+func onMessageHandler(msg *C.received_message, udata unsafe.Pointer) {
+	messageList := (gopointer.Restore(udata)).(*list.List)
+
+	topic := whispertypes.TopicType{}
+	copy(topic[:], C.GoBytes(unsafe.Pointer(&msg.topic[0]), whispertypes.TopicLength)[:whispertypes.TopicLength])
+	wrappedMsg := &whispertypes.Message{
+		TTL:       uint32(msg.ttl),
+		Timestamp: uint32(msg.timestamp),
+		Topic:     topic,
+		Payload:   C.GoBytes(unsafe.Pointer(msg.decoded), C.int(msg.decodedLen)),
+		PoW:       float64(msg.pow),
+		Hash:      C.GoBytes(unsafe.Pointer(&msg.hash[0]), statusproto.HashLength),
+		P2P:       true,
+	}
+	if msg.source != nil {
+		wrappedMsg.Sig = append([]byte{0x04}, C.GoBytes(unsafe.Pointer(msg.source), whispertypes.PubKeyLength)...)
+	}
+	if msg.recipientPublicKey != nil {
+		wrappedMsg.Dst = append([]byte{0x04}, C.GoBytes(unsafe.Pointer(msg.recipientPublicKey), whispertypes.PubKeyLength)...)
+	}
+
+	messageList.PushBack(wrappedMsg)
+}
+
+func (w *nimbusWhisperWrapper) Subscribe(privateKeyID, symKeyID string, pow float64, topics [][]byte) (string, error) {
+	f, err := w.createFilterWrapper("", privateKeyID, symKeyID, pow, topics)
+	if err != nil {
+		return "", err
+	}
+
+	retVal := w.routineQueue.Send(func(c chan<- interface{}) {
+		// Create a message store for this filter, so we can add new messages to it from the nimbus_subscribe_filter callback
+		messageList := list.New()
+		idC := C.nimbus_subscribe_filter(GetNimbusFilterFrom(f), (C.received_msg_handler)(unsafe.Pointer(C.onMessageHandler_cgo)), gopointer.Save(messageList))
+		filterID := C.GoString(idC)
+		if filterID == "" {
+			c <- errors.New("failed to subscribe to filter in Nimbus")
+		}
+
+		w.filterMessages[filterID] = messageList // TODO: Check if this is done too late (race condition with onMessageHandler)
+
+		f.(*nimbusFilterWrapper).id = filterID
+
+		c <- filterID
+	})
+	if err, ok := retVal.(error); ok {
+		return "", err
+	}
+
+	return retVal.(string), nil
+}
+
+func (w *nimbusWhisperWrapper) GetFilter(id string) whispertypes.Filter {
+	idC := C.CString(id)
+	defer C.free(unsafe.Pointer(idC))
+
+	panic("GetFilter not implemented")
+	// pFilter := C.nimbus_get_filter(idC)
+	// return NewNimbusFilterWrapper(pFilter, id, false)
+}
+
+func (w *nimbusWhisperWrapper) Unsubscribe(id string) error {
+	return w.routineQueue.Send(func(c chan<- interface{}) {
+		idC := C.CString(id)
+		defer C.free(unsafe.Pointer(idC))
+
+		if ok := C.nimbus_unsubscribe_filter(idC); !ok {
+			c <- errors.New("filter not found")
```

Comment:
No return after sending an error? In this branch, an error and later `nil` will be sent to the channel.

Reply:
Oops, good catch! I wish Go had some kind of language mechanism to make code like this less repetitive (such as aspect-oriented programming support or AST macros).